### ISSUE TITLE
[BUG] SKPath boolean operation union

### ISSUE THREAD

**Issue by filippo-springa** [mono/SkiaSharp#2325]
**Description**

When I try to get an offset path from a source path with GetFillPath() and boolean operation union. I get different results if the path is at a different position.

As you can see the paths are the same but the only difference is the first point after the SVG m comand.
**Code**

var pathBug = SKPath.ParseSvgPathData("m 450.35434,822.38865 c 0.617,-0.3014 0.617,-1.18 -8e-4,-1.4897 l -88.8175,-0.01 v -14.1176 l 88.8175,0.047 c 0.6178,-0.2555 0.6178,-1.13424 8e-4,-1.43545 -8.9949,-1.11499 -17.90118,-2.85449 -26.6546,-5.20599 -16.69,-4.07019 -43.8432,-25.11432 -43.8432,-25.11432 l 104.71302,26.80432 c 0,0 -31.0553,32.1922 -34.21524,20.52164 z");

var pathCorrect = SKPath.ParseSvgPathData("m 602.35434,822.38865 c 0.617,-0.3014 0.617,-1.18 -8e-4,-1.4897 l -88.8175,-0.01 v -14.1176 l 88.8175,0.047 c 0.6178,-0.2555 0.6178,-1.13424 8e-4,-1.43545 -8.9949,-1.11499 -17.90118,-2.85449 -26.6546,-5.20599 -16.69,-4.07019 -43.8432,-25.11432 -43.8432,-25.11432 l 104.71302,26.80432 c 0,0 -31.0553,32.1922 -34.21524,20.52164 z");
	
	
	var p = new SKPaint
            {
                Color = SKColors.Blue,
                StrokeJoin = SKStrokeJoin.Round,
                Style = SKPaintStyle.Stroke,
                IsAntialias = true,
                StrokeCap = SKStrokeCap.Round,
                StrokeWidth = 3.6f
            };
            
    var strokePath = new SKPath();
    p.GetFillPath(pathCorrect, strokePath, 2);
    
    var offsetPath = new SKPath();
    pathCorrect.Op(strokePath, SKPathOp.Union, offsetPath);
    
    var strokePathBug = new SKPath();
    p.GetFillPath(pathBug, strokePathBug, 2);
    
    var offsetPathBug = new SKPath();
    pathBug.Op(strokePathBug, SKPathOp.Union, offsetPathBug);

	p.StrokeWidth = 1;
	
	p.Color = SKColors.Red;
	canvas.DrawPath(pathCorrect, p);
	canvas.DrawPath(pathBug, p);
	
	p.Color = SKColors.Green;
	canvas.DrawPath(offsetPath, p);
	canvas.DrawPath(offsetPathBug, p);
**Actual Behavior**

![image](https://user-images.githubusercontent.com/117271075/204579760-ad11fc86-caac-4c4a-bea3-1b07b65568cc.png)

the result on the left should looks like to the one on the right

**Basic Information**

- Version with issue: 2.88.3 
- IDE:  Visual Studio 
- Platform Target Frameworks: 
- Windows Classic: 10, 11

**Comment by themcoo** [comment]
Hi,
this is rather a google/skia bug, not SkiaSharp... 
I found more issues with PathOps during my work, some were already pointed out in chromium backlog:
https://bugs.chromium.org/p/skia/issues/detail?id=10366&q=Area%3DPathOps&can=1&sort=-id

I also found out that they want to discourage clients from using PathOps: https://bugs.chromium.org/p/skia/issues/detail?id=10978&q=Area%3DPathOps&can=1&sort=-id

From my experience the PahtOps for paths containing Conics were always somewhat buggy. At some point I built a conic-remover for paths, which uses quads to approximate the conic.
```

        SKPath RemoveConics(SKPath path)
	{
		var result = new SKPath();
		var iter = path.CreateRawIterator();
		var pts = new SKPoint[4];
		SKPathVerb verb;
		while ((verb = iter.Next(pts)) != SKPathVerb.Done)
		{
			switch (verb)
			{
				case SKPathVerb.Conic:
					var quads = SKPath.ConvertConicToQuads(pts[0], pts[1], pts[2], iter.ConicWeight(), 2);
					for (var qi = 0; qi < quads.Length - 2; qi += 2)
					{
						result.QuadTo(quads[qi], quads[qi + 1]);
					}
					break;
				case SKPathVerb.Cubic:
					result.CubicTo(pts[1], pts[2], pts[3]);
					break;
				case SKPathVerb.Move:
					result.MoveTo(pts[0]);
					break;
				case SKPathVerb.Close:
					result.Close();
					break;
				case SKPathVerb.Done:
					return result;
				case SKPathVerb.Line:
					result.LineTo(pts[1]);
					break;
				case SKPathVerb.Quad:
					result.QuadTo(pts[1], pts[2]);
					break;
			}
		}

		return result;
	}

```
It does indeed fix your case:
```
	var strokePathBug = new SKPath();
	p.GetFillPath(pathBug, strokePathBug, 2);
        strokePathBug = RemoveConics(strokePathBug);
```
![image](https://user-images.githubusercontent.com/2964157/204833182-77c48a89-a5c4-42ba-a457-e5984fc53631.png)

**Comment by filippo-springa** [comment]
Hi,

Thanks for the fast reply!

I understand that is a google/skia bug.

However it didn't work on my program

**Comment by themcoo** [comment]
Hi,

sorry I made a mistake when rewriting the simplified code, the `for` loop should start from 1:
`    for (var qi = 1; qi < quads.Length - 2; qi += 2)`
(I edited original code). The first point is always the same as the last point from the previews verb.

If this also won't work for you, let me know what's the path that still causes issues. I wonder too!

And regarding your question, that's funny because it has been actually in my backlog forever and I haven't had have time to explore this yet, you kinda mobilized me to do so :) For path operations I'm using clipper library written (among others) in C#: http://www.angusj.com/clipper2/Docs/Overview.htm
Apparently a major update 2.0 has been released, I need to explore this. I was using version 1, I don't know whether the same limitations are still present in 2.0, but main thing is that it only works on polygons, so you need to flatten all curves before processing them with this library.

**Comment by filippo-springa** [comment]
Hi,

I tried with that fix but I had no luck. I'm using this path:

"m 450.35434,822.38865 c 0.617,-0.3014 0.617,-1.18 -8e-4,-1.4897 l -88.8175,-0.01 v -14.1176 l 88.8175,0.047 c 0.6178,-0.2555 0.6178,-1.13424 8e-4,-1.43545 -8.9949,-1.11499 -17.90118,-2.85449 -26.6546,-5.20599 -16.69,-4.07019 -43.8432,-25.11432 -43.8432,-25.11432 l 104.71302,26.80432 c 0,0 -31.0553,32.1922 -34.21524,20.52164 z"

I had a little improvement with your fix with the top left corner but the main issue remains.
If this works on your code maybe I'm doing something wrong :)

Anyway It appears that the problem occurs only in some section of the canvas(this image is without your workaround):

![image](https://user-images.githubusercontent.com/117271075/205010954-957e1869-e866-4760-a767-74b8ffe66cf2.png)

I have also opened an issue on skia/google if you want to follow up: https://bugs.chromium.org/p/skia/issues/detail?id=13954

Thanks for sharing me that clipper library, I will explore it and see if it fits for me!
